receiver/githubreceiver/logs.go

```diff
@@ -0,0 +1,315 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package githubreceiver // import "github.com/observiq/bindplane-agent/receiver/github"
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+var gitHubMaxLimit = 100
+
+
+
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+	CloseIdleConnections()
+}
+
+// githubLogsReceiver is a receiver for GitHub logs.
+type gitHubLogsReceiver struct {
+	client   httpClient
+	logger   *zap.Logger
+	consumer consumer.Logs
+	cfg      *Config
+	wg       *sync.WaitGroup
+	cancel   context.CancelFunc
+	nextURL  string
+}
+
+// newGitHubLogsReceiver creates a new GitHub logs receiver.
+func newGitHubLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) (*gitHubLogsReceiver, error) {
+	return &gitHubLogsReceiver{
+		cfg:      cfg,
+		logger:   logger,
+		consumer: consumer,
+		wg:       &sync.WaitGroup{},
+		client:   http.DefaultClient,
+	}, nil
+}
+
+// Start begins the receiver's operation.
+func (r *gitHubLogsReceiver) Start(_ context.Context, _ component.Host) error {
+	r.logger.Info("Starting GitHub logs receiver")
+	ctx, cancel := context.WithCancel(context.Background())
+	r.cancel = cancel
+	r.wg.Add(1)
+	if r.cfg.WebhookConfig != nil {
+		// Set up webhook handling
+		go r.setupWebhook()
+	}
+	if r.cfg.PollInterval > 0 {
+		// Start polling
+		go r.startPolling(ctx)
+	}
+	return nil
+
+}
+
+func (r *gitHubLogsReceiver) setupWebhook() error {
+	// waiting on open source code for webhooks
+	return nil
+}
+
+func (r *gitHubLogsReceiver) startPolling(ctx context.Context) {
+	defer r.wg.Done()
+	t := time.NewTicker(r.cfg.PollInterval)
+	err := r.poll(ctx)
+	if err != nil {
+		r.logger.Error("there was an error during the first poll", zap.Error(err))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			err := r.poll(ctx)
+			if err != nil {
+				r.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *gitHubLogsReceiver) poll(ctx context.Context) error {
+	logEvents := r.getLogs(ctx)
+	observedTime := pcommon.NewTimestampFromTime(time.Now())
+	logs := r.processLogEvents(observedTime, logEvents)
+	if logs.LogRecordCount() > 0 {
+		if err := r.consumer.ConsumeLogs(ctx, logs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *gitHubLogsReceiver) getLogs(ctx context.Context) ([]gitHubEnterpriseLog) {
```

Comment:
nit: you don't need to add parenthesis around the return type if you are returning just 1 thing.